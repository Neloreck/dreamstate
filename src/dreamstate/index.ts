/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export {
  ILoadable as Loadable,
  TNested as Nested,
  TComputed as Computed,
  TStateSetter as StateSetter,
  TConsumable as Consumable,
  ISignalEvent as SignalEvent,
  ISignal as Signal,
  TSignalType as SignalType,
  TQueryResponse as QueryResponse,
  TOptionalQueryResponse as OptionalQueryResponse,
  TQueryRequest as QueryRequest,
  TQueryType as QueryType
} from "@/dreamstate/types";

export { ContextService } from "@/dreamstate/core/services/ContextService";
export { ContextManager } from "@/dreamstate/core/services/ContextManager";

export { getCurrentContext } from "@/dreamstate/core/registry/getCurrentContext";
export { getCurrent } from "@/dreamstate/core/registry/getCurrent";

export { createProvider } from "@/dreamstate/core/provision/createProvider";
export { Provide } from "@/dreamstate/core/provision/Provide";
export { withProvision } from "@/dreamstate/core/provision/withProvision";

export { OnSignal } from "@/dreamstate/core/signals/OnSignal";
export { useSignals } from "@/dreamstate/core/signals/useSignals";
export { unsubscribeFromSignals } from "@/dreamstate/core/signals/unsubscribeFromSignals";
export { subscribeToSignals } from "@/dreamstate/core/signals/subscribeToSignals";
export { emitSignal } from "@/dreamstate/core/signals/emitSignal";

export { OnQuery } from "@/dreamstate/core/queries/OnQuery";
export { queryData } from "@/dreamstate/core/queries/queryData";

export { useManager } from "@/dreamstate/core/consumption/useManager";
export { Consume } from "@/dreamstate/core/consumption/Consume";
export { withConsumption } from "@/dreamstate/core/consumption/withConsumption";

export { Bind } from "@/dreamstate/utils/Bind";
export { createLoadable } from "@/dreamstate/utils/createLoadable";
export { createNested } from "@/dreamstate/utils/createNested";
export { createComputed } from "@/dreamstate/utils/createComputed";
export { createSetter } from "@/dreamstate/utils/createSetter";
