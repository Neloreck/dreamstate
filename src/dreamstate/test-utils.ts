/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export { ScopeContext as InternalScopeContext } from "@/dreamstate/core/scoping/ScopeContext";

export * from "@/dreamstate/test-utils/registry/getCurrent";
export * from "@/dreamstate/test-utils/utils/getReactConsumer";
export * from "@/dreamstate/test-utils/utils/getReactProvider";

export * from "@/dreamstate/test-utils/registry/mockManagerWithScope";
export * from "@/dreamstate/test-utils/registry/mockScopeProvider";
export * from "@/dreamstate/test-utils/registry/mockScope";

export * from "@/dreamstate/test-utils/utils/nextAsyncQueue";
