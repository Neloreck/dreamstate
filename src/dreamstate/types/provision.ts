import type { ReactNode } from "react";

/**
 * Properties for the Dreamstate context manager provider.
 * Defines the props used for initializing and configuring the provider.
 */
export interface IProviderProps<T> {
  initialState?: T;
  partialHotReplacement?: boolean;
  children?: ReactNode;
}
