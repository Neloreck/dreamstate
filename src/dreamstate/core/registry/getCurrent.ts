import { CONTEXT_SERVICES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyObject, TDreamstateService } from "@/dreamstate/types";

/**
 * Get current manager instance from registry.
 * Returns null if nothing is found.
 */
export function getCurrent<T extends TDreamstateService<any>>(
  Service: T
): InstanceType<T> | null {
  return CONTEXT_SERVICES_REGISTRY.get(Service) as InstanceType<T> || null;
}
