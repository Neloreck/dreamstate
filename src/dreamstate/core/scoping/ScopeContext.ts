import { Context, createContext } from "react";

import { IRegistry } from "@/dreamstate/core/scoping/registry/createRegistry";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import {
  IBaseSignal,
  IContextManagerConstructor,
  IOptionalQueryRequest,
  ISignalEvent,
  TAnyContextManagerConstructor,
  TAnyObject,
  TAnyValue,
  TCallable,
  TQueryListener,
  TQueryResponse,
  TQueryType,
  TSignalListener,
  TUpdateObserver,
  TUpdateSubscriber,
} from "@/dreamstate/types";

/**
 * Internals of the scope context, meant for internal use only.
 *
 * This interface defines the internal properties and methods for managing data flow, signals,
 * and queries within the scope. It is not intended for external access and should only be used
 * by the `dreamstate` library for its internal operations.
 */
export interface IScopeContextInternals {
  /**
   * Mutable registry that contains all references and scope data.
   */
  REGISTRY: IRegistry;

  /**
   * Registers a `ManagerClass` in the current scope and creates an instance for context provision.
   *
   * @template T - Type of initial state injected into the manager.
   * @template C - Type of initial context injected into the manager.
   * @param {IContextManagerConstructor} ManagerClass - The manager class that should be registered in the scope.
   * @param {T | null} [initialState] - Optional initial state parameter that will be injected into the manager's
   *   constructor upon creation.
   * @param {Partial<C>} [initialContext] - Optional initial context parameter that will be injected into the
   *   manager and mixed with the context.
   * @returns {boolean} `true` if the service was successfully registered, `false` otherwise.
   */
  registerService<T extends TAnyObject, C extends TAnyObject>(
    ManagerClass: TAnyContextManagerConstructor,
    initialState?: T | null,
    initialContext?: Partial<C>
  ): boolean;

  /**
   * Unregisters a `ManagerClass` from the current scope.
   * Cleans up memory and removes all internal references associated with the manager.
   *
   * @param {TAnyContextManagerConstructor} ManagerClass - The manager class that should be unregistered from the scope.
   * @returns {boolean} `true` if the service was successfully unregistered, `false` otherwise.
   */
  unRegisterService(ManagerClass: TAnyContextManagerConstructor): boolean;

  /**
   * Adds an observer to a specified service.
   * The observer will be notified when the manager's context is updated.
   *
   * @template T - Type of the manager's context.
   * @param {TAnyContextManagerConstructor} ManagerClass - The manager class reference for which the observer
   *   should be added.
   * @param {TUpdateObserver} serviceObserver - The observer that should be notified on manager updates.
   * @param {number} [referencesCount] - Optional count of manager references for injection.
   *   By default, checks the registry.
   */
  addServiceObserver(
    ManagerClass: TAnyContextManagerConstructor,
    serviceObserver: TUpdateObserver,
    referencesCount?: number
  ): void;

  /**
   * Removes an observer from a specified service.
   * The observer will no longer be notified of updates from the manager's context.
   *
   * @template T - Type of the manager's context.
   * @param {TAnyContextManagerConstructor} ManagerClass - The manager class reference for which
   *   the observer should be removed.
   * @param {TUpdateObserver} serviceObserver - The observer that should be removed from the registry.
   * @param {number} [referencesCount] - Optional count of manager references for injection.
   *   By default, checks the registry.
   */
  removeServiceObserver(
    ManagerClass: TAnyContextManagerConstructor,
    serviceObserver: TUpdateObserver,
    referencesCount?: number
  ): void;

  /**
   * Notifies all observers that are subscribed to a manager's context updates.
   * This will update all providers based on the current context of the provided manager instance.
   *
   * @template T - Type of the manager's context.
   * @param {ContextManager<T>} manager - The manager instance whose context should trigger updates
   *   to subscribed providers.
   */
  notifyObservers<T extends TAnyObject>(manager: ContextManager<T>): void;

  /**
   * Subscribes to manager context updates.
   * The provided callback will be triggered whenever the manager's context is updated.
   *
   * @template T - Type of the manager's context.
   * @template D - Type of the context manager constructor.
   * @param {IContextManagerConstructor} ManagerClass - The manager class reference for which to subscribe.
   * @param {TUpdateSubscriber<T>} subscriber - The callback function that will be triggered on context updates.
   * @returns {TCallable} A function that can be used to unsubscribe from context updates.
   */
  subscribeToManager<T extends TAnyObject, D extends IContextManagerConstructor<TAnyValue, T>>(
    ManagerClass: D,
    subscriber: TUpdateSubscriber<T>
  ): TCallable;

  /**
   * Unsubscribes from manager context updates.
   *
   * @template T - Type of the manager's context.
   * @template D - Type of the context manager constructor.
   * @param {IContextManagerConstructor} ManagerClass - The manager class reference for which to unsubscribe.
   * @param {TUpdateSubscriber<T>} subscriber - The callback function that should be removed
   *   from context updates subscription.
   */
  unsubscribeFromManager<T extends TAnyObject, D extends IContextManagerConstructor<TAnyValue, T>>(
    ManagerClass: D,
    subscriber: TUpdateSubscriber<T>
  ): void;
}

/**
 * Represents the current scope context.
 * Provides public methods to interact with the current scope, including managing context, signals, and queries.
 */
export interface IScopeContext {
  /**
   * Library internals.
   * Not intended for normal use, except during unit testing.
   */
  INTERNAL: IScopeContextInternals;

  /**
   * Retrieves the current context snapshot for the given manager class.
   *
   * @template T - Type of the context.
   * @template D - Type of the context manager constructor.
   * @param manager - The manager class whose context is to be retrieved.
   * @returns The current context snapshot for the specified manager.
   */
  getContextOf<T extends TAnyObject, D extends IContextManagerConstructor<T>>(manager: D): T;

  /**
   * Emits a signal for all subscribers in the current Dreamstate scope.
   *
   * @template D - Type of the signal data.
   * @param base - The base signal object that triggers the event.
   *   It includes a `type` field  and an optional `data` field.
   * @param emitter - Optional signal emitter reference.
   * @returns A signal event instance that wraps the emitted signal.
   */
  emitSignal<D = undefined>(base: IBaseSignal<D>, emitter?: TAnyContextManagerConstructor | null): ISignalEvent<D>;

  /**
   * Subscribes to signals within the current scope.
   * The listener will be invoked on every signal event.
   *
   * @template D - Type of the signal data.
   * @param listener - The callback function to be invoked on signal events.
   * @returns A callable function that unsubscribes the listener.
   */
  subscribeToSignals<D = undefined>(listener: TSignalListener<D>): TCallable;

  /**
   * Unsubscribes a listener from signals in the current scope.
   * The provided listener will no longer be invoked on signal events.
   *
   * @template D - Type of the signal data.
   * @param listener - The callback function to remove from signal subscriptions.
   */
  unsubscribeFromSignals<D = undefined>(listener: TSignalListener<D>): void;

  /**
   * Registers a query provider callback to answer query data calls.
   *
   * @template T - Type of the query.
   * @param queryType - The type of query for data provisioning.
   * @param listener - The callback that listens to queries and returns evaluation data.
   * @returns A callable function that unregisters the query provider.
   */

  registerQueryProvider<T extends TQueryType>(queryType: T, listener: TQueryListener<T, TAnyValue>): TCallable;

  /**
   * Unregisters a query provider callback.
   * The callback will no longer handle query data calls.
   *
   * @template T - Type of the query.
   * @param queryType - The type of query for which the provider is unregistered.
   * @param listener - The callback to be unregistered.
   */
  unRegisterQueryProvider<T extends TQueryType>(queryType: T, listener: TQueryListener<T, TAnyValue>): void;

  /**
   * Queries data synchronously from the current scope.
   * The handler for the specified query type is executed, and its result is wrapped as a query response.
   *
   * @template D - Type of the data provided in the query request.
   * @template T - Type of the query.
   * @template Q - Type of the query request.
   * @param query - The query request containing the query type and optional data.
   * @returns The query response or null if no handler is found.
   */
  queryDataSync<D, T extends TQueryType, Q extends IOptionalQueryRequest<D, T>>(query: Q): TQueryResponse<TAnyValue>;

  /**
   * Queries data asynchronously from the current scope.
   * The handler for the specified query type is executed, and its result is wrapped as a query response.
   *
   * @template D - Type of the data provided in the query request.
   * @template T - Type of the query.
   * @template Q - Type of the query request.
   * @param query - The query request containing the query type and optional data.
   * @returns A promise that resolves with the query response or null if no handler is found.
   */
  queryDataAsync<D, T extends TQueryType, Q extends IOptionalQueryRequest<D, T>>(
    query: Q
  ): Promise<TQueryResponse<TAnyValue>>;
}

/**
 * React context reference for the Dreamstate scope, used to wrap the virtual DOM tree.
 * This context allows the React tree to be aware of the current scope and provides access
 * to context management functionalities within the Dreamstate data flow.
 */
export const ScopeContext: Context<IScopeContext> = createContext(null as unknown as IScopeContext);
