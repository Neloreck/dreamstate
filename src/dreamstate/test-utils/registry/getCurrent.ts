import { IPublicScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Get current manager instance from current scope.
 * Returns null if nothing is found.
 */
export function getCurrent<T extends TAnyContextManagerConstructor>(
  Service: T,
  scope: IPublicScopeContext
): InstanceType<T> | null {
  return scope.REGISTRY.CONTEXT_SERVICES_REGISTRY.get(Service) as InstanceType<T>|| null;
}
