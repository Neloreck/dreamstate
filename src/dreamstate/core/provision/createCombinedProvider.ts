import { FunctionComponent, ReactNode } from "react";

import { IProviderProps } from "@/dreamstate/core/provision/createProvider";
import { provideSubTreeRecursive } from "@/dreamstate/core/provision/provideSubTreeRecursive";
import { useSourceObserving } from "@/dreamstate/core/provision/useSourceObserving";
import { TAnyContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Create provider that unifies all data sources and tries to re-render providers on every update.
 */
export function createCombinedProvider<T extends TAnyObject>(
  sources: Array<TAnyContextManagerConstructor>
): FunctionComponent<IProviderProps<T>> {
  function Observer(props: IProviderProps<T>): ReactNode {
    const registry: Map<TAnyContextManagerConstructor, TAnyObject> = useSourceObserving(sources, props);

    return provideSubTreeRecursive(props.children, sources, registry);
  }

  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextManagerConstructor) {
      return it.name;
    })
    }]`;
  } else {
    Observer.displayName = "DS.Observer";
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
