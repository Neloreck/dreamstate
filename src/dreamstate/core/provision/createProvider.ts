import { FunctionComponent, ReactNode, useCallback, useState } from "react";

import { EMPTY_ARR } from "@/dreamstate/core/internals";
import { useHotObservers } from "@/dreamstate/core/observing/useHotObservers";
import { useStaticObservers } from "@/dreamstate/core/observing/useStaticObservers";
import { provideSubTreeRecursive } from "@/dreamstate/core/provision/provideSubTreeRecursive";
import { ContextManager } from "@/dreamstate/core/services/ContextManager";
import { ContextService } from "@/dreamstate/core/services/ContextService";
import {
  TAnyContextManagerConstructor,
  TAnyContextServiceConstructor,
  TAnyObject
} from "@/dreamstate/types";

export interface IProviderProps<T> {
  initialState?: T;
  partialHotReplacement?: boolean;
  children?: ReactNode;
}

/**
 * Create component for manual provision without HOC/Decorator-like api.
 * Useful if your root is functional component or you are using createComponent api without JSX.
 */
export function createProvider<
  T extends TAnyObject = TAnyObject
>(
  sources: Array<TAnyContextServiceConstructor>
): FunctionComponent<IProviderProps<T>> {
  if (!Array.isArray(sources)) {
    throw new TypeError(
      "Wrong providers parameter supplied. Only array of context services is acceptable."
    );
  }

  for (let it = 0; it < sources.length; it ++) {
    if (!sources[it] || !(sources[it].prototype instanceof ContextService)) {
      throw new TypeError("Only classes extending ContextService can be supplied for provision.");
    }
  }

  /**
   * Check only managers with required provision.
   * Do not include services for subTree rendering but add registering logic for services.
   */
  const managers: Array<TAnyContextManagerConstructor> = sources.filter(function(
    Service: TAnyContextServiceConstructor
  ) {
    return Service.prototype instanceof ContextManager;
  }) as Array<TAnyContextManagerConstructor>;

  /**
   * Create observer component that will handle observing.
   */
  function Observer(props: IProviderProps<T>): ReactNode {
    const [ , forceRender ] = useState({});
    const updateProviders = useCallback(function() {
      forceRender({});
    }, EMPTY_ARR);

    if (props.partialHotReplacement) {
      useHotObservers(sources, props.initialState, updateProviders);
    } else {
      useStaticObservers(sources, props.initialState, updateProviders);
    }

    return provideSubTreeRecursive(props.children, managers, 0);
  }

  if (IS_DEV) {
    Observer.displayName = `Dreamstate.Observer[${sources.map(function(it: TAnyContextServiceConstructor) {
      return it.name;
    })
    }]`;
  } else {
    Observer.displayName = "DS.Observer";
  }

  return Observer as FunctionComponent<IProviderProps<T>>;
}
