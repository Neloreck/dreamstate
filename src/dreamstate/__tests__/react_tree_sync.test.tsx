import { mount } from "enzyme";
import { default as React, ReactElement } from "react";
import { act } from "react-dom/test-utils";

import { ScopeProvider, useManager, useScope } from "@/dreamstate";
import { createProvider } from "@/dreamstate/core/provision/createProvider";
import { IScopeContext } from "@/dreamstate/core/scoping/ScopeContext";
import { getCurrent } from "@/dreamstate/test-utils/registry/getCurrent";
import { ITestContext, TestManager } from "@/fixtures";

describe("React tree for provided and consumed components", () => {
  it("Should correctly update subscribed to functional provider elements view", () => {
    const ContextProvider = createProvider([ TestManager ]);

    function Root(): ReactElement {
      return (
        <ScopeProvider>
          <ContextProvider>
            <Consumer/>
          </ContextProvider>
        </ScopeProvider>
      );
    }

    let testContextManager: TestManager | null = null;

    function Consumer(): ReactElement {
      const scope: IScopeContext = useScope();
      const context: ITestContext = useManager(TestManager);

      testContextManager = getCurrent(TestManager, scope);

      return <div> { JSON.stringify(context) }</div>;
    }

    const reactTree = mount(<Root/>);

    expect(reactTree.render()).toMatchSnapshot();

    act(() => {
      testContextManager!.setContext({ first: "updated", second: 101010, third: true });
    });

    expect(testContextManager!.context.first).toBe("updated");
    expect(testContextManager!.context.second).toBe(101010);
    expect(testContextManager!.context.third).toBe(true);

    reactTree.update();
    expect(reactTree.render()).toMatchSnapshot();

    reactTree.unmount();
  });
});
