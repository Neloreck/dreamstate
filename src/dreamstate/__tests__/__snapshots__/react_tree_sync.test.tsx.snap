// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React tree for provided and consumed components Should correctly update subscribed to functional provider elements view 1`] = `
<div>
   {"first":"first","second":2,"third":false}
</div>
`;

exports[`React tree for provided and consumed components Should correctly update subscribed to functional provider elements view 2`] = `
<div>
   {"first":"updated","second":101010,"third":true}
</div>
`;
